core/Engine/Engine.Drivers/Context/InputValidationContextDriver.cs

```diff
@@ -0,0 +1,166 @@
+using System.Threading.Tasks;
+using Engine.DataTypes;
+using FSharpUtils.Newtonsoft;
+using LanguageExt;
+using Microsoft.FSharp.Core;
+using Newtonsoft.Json;
+using Tweek.Utils;
+using static System.Tuple;
+using static LanguageExt.Prelude;
+using Microsoft.Extensions.Logging;
+
+namespace Engine.Drivers.Context
+{
+    public class InputValidationContextDriver : IContextDriver
+    {
+        public enum Mode
+        {
+            AllowUndefinedProperties,
+            Strict,
+            ReportOnly
+        }
+        public static readonly EventId ValidationFailedEventId = new EventId(6754, "ContextInputValidationFailed");
+        public delegate Option<JsonValue> IdentitySchemaProvider(string identityType);
+        public delegate Option<CustomTypeDefinition> CustomTypeDefinitionProvider(string typeName);
+
+        private readonly IContextDriver _child;
+        private readonly ILogger<InputValidationContextDriver> _logger;
+        private readonly IdentitySchemaProvider _identitySchemaProvider;
+        private readonly CustomTypeDefinitionProvider _customTypeDefinitionProvider;
+        private readonly Mode _mode;
+        private readonly bool _reportOnly;
+
+        public InputValidationContextDriver(
+            IContextDriver child, 
+            ILogger<InputValidationContextDriver> logger,
+            IdentitySchemaProvider identitySchemaProvider, 
+            CustomTypeDefinitionProvider customTypeDefinitionProvider, 
+            Mode mode = Mode.Strict,
+            bool reportOnly = false)
+        {
+            this._child = child;
+            this._logger = logger;
+            this._identitySchemaProvider = identitySchemaProvider;
+            this._customTypeDefinitionProvider = customTypeDefinitionProvider;
+            this._mode = mode;
+            this._reportOnly = reportOnly;
+        }
+
+        public Task AppendContext(Identity identity, Dictionary<string, JsonValue> context)
+        {
+            var result = 
+                _identitySchemaProvider(identity.Type)
+                .Map(schema => 
+                    context.Where(prop => !prop.Key.StartsWith("@fixed:"))
+                        .Select(item => ValidateSingleProperty(item, schema))
+                        .Aggregate(Create(true, ""), (aggregation, next) => next.Item1 ? aggregation : next)
+                )
+                .IfNone(Create(false, $"schema for identity type \"{identity.Type}\" not found"));
+
+            if (!result.Item1)
+            {
+                if (!_reportOnly)
+                {
+                    throw new ArgumentException(result.Item2);
+                }
+                else {
+                    _logger.LogWarning(result.Item2);
+                }
+            }
+            
+            return _child.AppendContext(identity, context);
+        }
+
+        private Tuple<bool, string> ValidateSingleProperty(KeyValuePair<string, JsonValue> item, JsonValue schema)
```

Comment:
you can return (bool isValid, string validationError), c# 7.0 has support for named tuples, and it'll make it clearer (instead of Item1, Item2), from a performance perspective, it's actually the same.

Reply:
Cool, wasn't aware of that. Huge improve to readability...